index.bs

```diff
@@ -626,14 +655,244 @@ interface MLContext {};
+                    1. The length of |resource| must be the same as |inputSize|.
+            1. For each |key| -> |value| of |outputs|:
+                1. |graph|.{{MLGraph/[[outputNames]]}}[|key|] must exist.
+        </div>
+    <!-- Compute -->
+    1. For each |key| -> |value| of |inputs|:
+        1. Let |inputDesc| be |graph|.{{MLGraph/[[inputDescriptors]]}}[|key|].
+        1. Let |inputTensor| be a new tensor for |graph|.{{MLGraph/[[implementation]]}} of data type that is compatible with |inputDesc|.{{MLOperandDescriptor/type}}.
+        1. If |value| is an {{MLInput}}, then:
+            1. Set the dimensions of |inputTensor| to |value|.{{MLInput/dimensions}}.
+        1. Else:
+            1. Set the dimensions of |inputTensor| to |inputDesc|.{{MLOperandDescriptor/dimensions}}.
+        1. If |value| is an {{MLInput}}, then:
+            1. Set the values of |inputTensor| to the values of |value|.{{MLInput/resource}}.
+        1. If |value| is an {{MLResource}}, then:
+            1. Set the values of |inputTensor| to the values of |value|.
+        1. Set the input of |graph|.{{MLGraph/[[implementation]]}} that is associated with |key| to |inputTensor|.
+    1. For each |key| -> |value| of |outputs|:
+        1. Issue a compute request for output of |graph|.{{MLGraph/[[implementation]]}} that is associated with |key|.
+        1. Wait for the compute request to be completed.
+        1. If there is an error returned by |graph|.{{MLGraph/[[implementation]]}}, then:
+            1. Throw an {{OperationError}} {{DOMException}} and stop.
+        1. Else:
+            1. Let |outputTensor| be the output tensor returned by |graph|.{{MLGraph/[[implementation]]}}.
+            1. If the kind of |value| is not compatible with the value type of |outputTensor|, then throw a {{DataError}} {{DOMException}} and stop.
+            1. Let |outputSize| be 1.
+            1. For each |dimension| of dimensions of |outputTensor|:
+                1. Set |outputSize| to the product of |outputSize| and |dimension|.
+            1. If |outputSize| is greater than the length of |value|, then:
+                1. Throw a {{DataError}} {{DOMException}} and stop.
+            1. Else:
+                1. Set the values of |value| to the values of |outputTensor|.
+    1. Return Promise<{{undefined}}>.
+</div>
+
+<div class="note">
+When the {{[[contextType]]}} of {{MLContext}} is set to [=webgpu-context|webgpu=] with the {{MLContextOptions}}.{{gpuDevice}} also specified, the user agent queues the ML workload needed to compute the result of the operations in the graph in an internal command buffer before submitting it to the {{gpuDevice}}'s default {{GPUQueue}}. The user agent then returns the call with a {{Promise}} that resolves once the queue finishes processing all the work submitted to it.
```

Comment:
It is unclear to me that who submits the internal command buffer to gpuDevice's default GPUQueue. And when to submit? I am also a bit confused by the usage of the returned promise. Is its usage same as the promise returned by WebGPU's [`GPUQueue::onSubmittedWorkDone`](https://www.w3.org/TR/webgpu/#dom-gpuqueue-onsubmittedworkdone)?

Reply:
At some point, I suspect `computeAsync` will probably need to also accept a sequence of `GPUCommandBuffers` (created by WebGPU, not WebNN) if WebNN is to efficiently interleave 3D/Compute and ML work. 

WebNN's implementation of `computeAsync` has no way to de-dup submits. `GPUQueue.Submit` was intended to be called _only once per frame_ - the per-submit call overhead by the WebGPU runtime is quite high due to GC. So long as `computeAsync` is responsible to _execute_ `GPUCommandBuffers`, there is no issue mixing ML and GPU commands in them and batching them together into a single `computeAsync` call will be far more efficient for interop. 

Hopefully that makes the issue a bit clearer.

Reply:
> At some point, I suspect computeAsync will probably need to also accept a sequence of GPUCommandBuffers 

At that point, I'd much rather reintroduce the `MLCommandEncoder` interface and have all the fine-grained, intra-command buffer interop be done there as opposed to overloading `computeAsync` to become something much more complicated. 

> WebNN's implementation of computeAsync has no way to de-dup submits. 

Given that `computeAsync` will submit to the queue on behalf of the caller the content that is only generated and recorded within its own internal command buffer, I don't see a possibility of `computeAsync` causing a double submission of the same workload if that's what you're referring to. Am I missing anything?

Reply:
@wchao1115

Consider what happens if a WebGPU buffer, being rendered into by WebGPU is also referenced by the internal command buffer generated by WebNN (eg. shared tensor, if you will). If the WebGPU developer does not call submit, the buffer contents will not be updated. Meanwhile, if `computeAsync` only "submits" it's own internal command buffer, it will see the wrong result. To remedy, the WebNN developer has no choice but to submit twice. Once through WebGPU, then again using WebNN.  Now, what if I want to do a GPU pre-pass AND post-pass with some ML work in-between? That's at-least 2x WebGPU submits per frame or "submit per ml context". If the cost of submission exceeds the benefit of using ML, no WebNN for you.

Reply:
> Now, what if I want to do a GPU pre-pass AND post-pass with some ML work in-between?

The [background blur sample](https://github.com/webmachinelearning/webnn/issues/226#issuecomment-1074968279) requires to execute the following steps for every video frame:

1. Import `VideoFrame` into a WebGPU `GPUTexture`.
2. Blur the imported texture by a WebGPU shader (blur.wgsl).
3. Prepare (normalize) the imported texture to the input tensor (`GPUBuffer`) by a WebGPU compute shader (preprocessing.wgsl).
4. Background segmentation by WebNN graph. Its input is the prepared `GPUBuffer` and its output is another `GPUBuffer` that contains the segmap.
6. Blend the imported texture and the blurred texture based on the segmap (blend.wgsl) into another `GPUTexture` (blend.wgsl). Then draw this texture into an offscreen canvas.
7. Export the offscreen canvas to a VideoFrame.

So with this design, there would be three submits per video frame: 1) `GPUQueue::submit` after step 3; 2) `MLContext::computeAsync` at step 4 submits the ML workload internally; 3) another `GPUQueue::submit` after step 6.

Do I understand correctly?

Reply:
@huningxin I believe you are correct. 

@bbernhar Correct. The input buffer to an ML graph requires the content to be properly in place before the graph is executed. It is the caller's responsibility to make sure that the content is first rendered before it being processed in the ML graph during an inference call. Technically, this can be done either by having both the render workload and the ML workload be in the same command buffer submitted together, or on separate command buffers submitted individually or together.  When you said "submit twice" I thought you meant having the same workload in a command buffer submitted twice. That is not the case, so thanks for clarifying it. 

Do we really think that the cost of multiple queue submissions can overshadow either the rendering cost or the inference cost in the ML graph? I find that hard to believe. So although having a single submission seems cleaner from the WebGPU point of view. I don't really think multiple submission is a deal breaker, especially if it buys us a clear and clean separation between how the rendering workload and ML workload are executed across WebGPU and WebNN. I think it's just a trade-off.

Reply:
@wchao1115 

I would argue MLCommandBuffer and GPUCommandBuffer, used by WebNN and WebGPU respectively, equally demonstrate a "clean separation". My point is submit has a _fixed CPU overhead_ and it's recommended (for WebGPU developers) to submit enough work (at once) to hide this latency. Otherwise, smaller submits finish faster then newer ones can be submitted and the GPU will sit idle.

Reply:
According to my experience of adding WebNN (segmentation) into a WebGPU-based (blur) pipeline, if the ML workload could be wrapped into an `MLCommandBuffer`, and be submitted together with pre and post `GPUCommandBuffer`s via a single `GPUQueue::submit`, it would be simple and clean.

I would also recommend to keep `computeAsync` just working with `ArrayBufferView`s. This would help separate the WebNN standalone usage from WebNN/WebGPU interop usage. I hope it would make it simple for developers to select the execution method based on the usage, e.g., something like:

1. If use WebNN API standalone, then:
   1. if need sync API, then:
      1. use `MLContext::compute` in worker thread
   1. else:
      1. use `MLContext::computeAsync`
3. else if use WebNN in WebGPU pipeline, then:
   1. (open to discuss) `MLCommandBuffer` or `MLCommandEncoder` proposal
5. else if use WebNN in WebGL pipeline, then:
   1. (need more investigation)

My points are also reflected in [the table](https://github.com/webmachinelearning/webnn/pull/257#issuecomment-1080120454) of earlier comment if that is more clear.

Reply:
I believe I've made my point in this PR that there is no perfect design that satisfies all requirements. And I feel we have been circling around what I see as an unavoidable trade-off between the two designs I've been proposing all along. 

1. WebNN returns a `GPUCommandBuffer` filled with ML workloads, then let the caller be responsible for all workload submission. Since the caller owns workload submission, they can fold and order graphics and ML workload in a single submission. This is what @huningxin is arguing for, and it is my original proposal. This design however poses an implementation challenge in that WebNN must also implement the returning command buffer such that the WebGPU implementation can understand and process. Simply because it requires that a WebNN method returns a WebGPU interface.
2. WebNN owns both the generation and the submission of its own ML workload via its own internal command buffer. The only dependency on WebGPU is the incoming `GPUDevice` and its associated default command queue. This is a simpler approach that decouples WebNN from any WebGPU-generated type (e.g. `GPUCommandBuffer` etc) but with a drawback that it may require multiple submissions from the caller in an interop scenario e.g. the blur prototype Ningxin is working on. It is also the design that I believe @bbernhar preferred in our offline conversation with @RafaelCintron on the topic before I removed `MLCommandEncoder` in the latest commit based on that consensus.

I hope I summarize it well enough here. Ultimately, unless someone else wants to propose something different, I think we need to choose between the two and agree on the trade-off that each design presents.

Reply:
@bbernhar Because `computeAsync` also has to take care of the CPU cases. Your proposal calls for the callers optionally pass in their `GPUCommandBuffer` with rendered content from WebGPU to the WebNN workload submission method. It's just a bit messy to have a method taking WebGPU command buffers while at the same time able to populate CPU ML work on a worker thread.

Reply:
@bbernhar 

> Not sure I understand why it would be more complicated to re-use MLContext::computeAsync for case (2) too, do you mind elaborating a bit more?

Because the `MLContext::computeAsync` also needs to support `ArrayBufferView` for case (1). If the compute method only submits ML workload to the device queue for case (2), it doesn't have to be async call, just like `GPUQueue::submit`. The returned promise seems duplicated and confusing to me. It also requires multiple queue submissions for video processing use case. And as @RafaelCintron [mentioned](https://github.com/webmachinelearning/webnn/issues/230#issuecomment-1031849403), if WebGPU adds multi-queue in the future, this method needs to be changed again since it currently only submits to the default queue. I suppose the complex comes by packing the two usages into one method. 

Another perspective is, if we keep `MLContext::computeAsync` (and `MLContext::compute`) only for case (1), it would give us a "clean separation" of the dependency. I think it would enable us move fast on the spec development for WebNN standalone usage (case (1)).

> Getting GPUQueue::submit to accept MLCommandBuffer (or GPUCommandBuffers filled with ML commands) requires a discussion with the WebGPU WG, at a minimum.

I would hope that conversation could be started as early as possible. There is an open [issue](https://github.com/gpuweb/gpuweb/issues/2500) in WebGPU repo for this. The current [plan](https://www.w3.org/2022/03/24-webmachinelearning-minutes.html#t07) is that we make the consensus first within Web ML WG and then we coordinate with WebGPU WG for broader review. Does the plan make sense?

Reply:
> If the compute method only submits ML workload to the device queue for case (2), it doesn't have to be async call, just like GPUQueue::submit

`GPUQueue::submit` is always `async`, hence my confusion of having another "async" execution path for WebNN (but ONLY for the CPU). It's not obvious to me it's a change of dispatch (ie. worker thread vs WARP) that distinguishes them, since the returned promise can be ignored.

Your plan SGTM.